src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
+    Tuple,
+    Unit,
+}
+
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct StructDef {
     /// Fields, not including ctor
     pub fields: Vec<StructField>,
     /// ID of the constructor. This is only used for tuple- or enum-like
     /// structs.
-    pub ctor_id: Option<NodeId>,
+    pub id: NodeId,
```

Comment:
@petrochenkov if the id is the id of the variant, then it makes sense to me for it to be on the variant struct, not the structdef struct, since that is used by both variants and struct items. In the struct item case, it seems the id is redundant unless it is used as a ctor id.